Mage.Sets/src/mage/cards/f/FireGiantsFury.java

```diff
@@ -0,0 +1,214 @@
+import mage.abilities.effects.common.continuous.GainAbilityTargetEffect;
+import mage.abilities.keyword.TrampleAbility;
+import mage.cards.Card;
+import mage.cards.CardImpl;
+import mage.cards.CardSetInfo;
+import mage.constants.*;
+import mage.filter.common.FilterControlledCreaturePermanent;
+import mage.game.Game;
+import mage.game.events.DamagedEvent;
+import mage.game.events.GameEvent;
+import mage.players.Player;
+import mage.target.TargetPermanent;
+import mage.target.targetpointer.FixedTarget;
+import mage.util.CardUtil;
+
+import java.util.Set;
+import java.util.UUID;
+
+/**
+ *
+ * @author weirddan455
+ */
+public final class FireGiantsFury extends CardImpl {
+
+    private static final FilterControlledCreaturePermanent filter = new FilterControlledCreaturePermanent(SubType.GIANT);
+
+    public FireGiantsFury(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.SORCERY}, "{1}{R}");
+
+        // Target Giant you control gets +2/+2 and gains trample until end of turn.
+        Effect effect = new BoostTargetEffect(2, 2, Duration.EndOfTurn);
+        effect.setText("Target Giant you control gets +2/+2");
+        this.getSpellAbility().addEffect(effect);
+
+        effect = new GainAbilityTargetEffect(TrampleAbility.getInstance(), Duration.EndOfTurn);
+        effect.setText("and gains trample until end of turn");
+        this.getSpellAbility().addEffect(effect);
+
+        // Whenever it deals combat damage to a player this turn, exile that many cards from the top of your library. Until the end of your next turn, you may play those cards.
+        effect = new FireGiantsFuryEffect();
+        this.getSpellAbility().addEffect(effect);
+
+        this.getSpellAbility().addTarget(new TargetPermanent(filter));
+    }
+
+    private FireGiantsFury(final FireGiantsFury card) {
+        super(card);
+    }
+
+    @Override
+    public FireGiantsFury copy() {
+        return new FireGiantsFury(this);
+    }
+}
+
+class FireGiantsFuryEffect extends OneShotEffect {
+
+    public FireGiantsFuryEffect() {
+        super(Outcome.PlayForFree);
+        this.staticText = "Whenever it deals combat damage to a player this turn, exile that many cards from the top of your library. Until the end of your next turn, you may play those cards";
+    }
+
+    private FireGiantsFuryEffect(final FireGiantsFuryEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public FireGiantsFuryEffect copy() {
+        return new FireGiantsFuryEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        DelayedTriggeredAbility delayedAbility = new FireGiantsFuryDelayedTriggeredAbility(source.getFirstTarget());
+        game.addDelayedTriggeredAbility(delayedAbility, source);
+        return true;
+    }
+}
+
+class FireGiantsFuryDelayedTriggeredAbility extends DelayedTriggeredAbility {
+
+    private final UUID target;
```

Comment:
this should be a `MageObjectReference` so you can track whether it's left the battlefield and returned, take a look at Hunter's Insight for reference.

Reply:
Yea, I was looking at Searing Blood and Reincarnation when I wrote this that use UUID like this in their delayed triggered abilities but this is probably better.  I changed it.

Reply:
There are definitely older cards that need to be refactored, it's an unending battle.